### ISSUE TITLE
Remove Constraint of uploading to a single Liason Storage Provider

### ISSUE THREAD

**Issue by mnaamani** [Joystream/joystream#2269]
We currently have a constraint in the data_directory that when a data object is created a storage provider is selected at random and assigned the responsibility of processing an upload. Only that provider can 'Accept' the content.

I propose to remove this constraint and any active provider should be able to process the upload, and the client uploading can at its own discretion select the "best" provider to upload to and will be able to re-try other providers on failure. Removing this constraint also means the client will never have to create a new data object when the liaison is unavailable for a long period of time, and when/if the liaison is fired, and not need to call into the content directory to perform channel or video Update operations just to re-attempt uploading assets and delete old assets.
This greatly improves the UX both from joystream-cli and Atlas viewpoint.

I've discussed this both with @bwhm and @kdembler and got positive feedback. The implementation changes in the runtime are pretty straight forward. This leaves the question of the impact on Atlas design changes, if any, will be required. As I understand from Klaudiusz the designers are currently in the process of working on this aspect of Atlas and now would be the best time to decide whether we should go ahead with this proposal.

cc: @bedeho

**Comment by bedeho** [comment]
I really like this, but it almost seems too good to be true. I wonder if we are overlooking something obvious.

Would there still be an on-chain confirmation step for a valid upload?

I think this is a good idea, if we are really sure, and it should be incorporated into v2

https://github.com/Joystream/joystream/issues/2224

This impacts some rather nuanced product related design considerations currently being worked on https://github.com/Joystream/atlas/issues/381. It may perhaps simplify them, but the work has already started along this path.

So in conclusion, **ACK** from me, but some some double checking and coordination with designers is needed.

**Comment by bwhm** [comment]
As implied by @mnaamani, also thinks this seems great for the testnet.

> I really like this, but it almost seems too good to be true. I wonder if we are overlooking something obvious.

If you're only looking at this for the testnet, I don't think so. On mainnet, there needs to be some way of reliably determining whether some SPs are not performing, but from previous conversations about the design of a new storage system, it seems like we are moving towards lots of off-chain monitoring`*` anyway.

`*` keeping tabs of the various actors (storage/bandwidth providers, gateway/query node operators) interacting in the "content system" are performing.

**Comment by mnaamani** [comment]
> Would there still be an on-chain confirmation step for a valid upload?

Yes that step will will still be there, and what I suggest is that the the data_object.liason field will updated to indicate which provider actually approved or rejected* the content. So we will at least have a record of which nodes are performing their job or not. But as Martin clearly points out there really is no reliable way on-chain to determine real performance of SPs.

* I do have some thoughts also about merits of actually allowing nodes to 'Reject' content on chain. it would be better for the status to remain pending which would allow uploader to retry with another node (in-case one nodes decides to be a bad actor and just reject all objects)

**Comment by mochet** [comment]
I brought up something similar in #2213.

Currently, we have a storage provider that generates DNS issues for some users and no issues for others, but if you are assigned this provider your upload will fail and there is no method I can see to retry or try another provider.

If we have hundreds of providers it could be that due to the nature of the internet that some amount of providers will be down for certain users all of the time, so from a reliability perspective it makes sense. It may also be that providers are blocked from the client-side by censorship, local router/spam/browser extension blacklists, ISP enforced blacklists, international routing issues and other situations (so the providers themselves may be 100% reliable, but they may not ever work for some clients)

**Comment by bedeho** [comment]
That issue did not catch my attention, apologies.

**Comment by mnaamani** [comment]
Implemented for sumer in #2278
